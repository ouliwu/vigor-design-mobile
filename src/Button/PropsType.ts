/**
 * Button properties
 */
export interface ButtonPropsType {
  /** 按钮样式 */
  type?: 'default' | 'primary' | 'warn';
  /** 是否禁用按钮 */
  disabled?: boolean;
  /** 按钮点触事件 */
  onPress?: (e?: React.SyntheticEvent) => void;
  /** 按钮是否显示加载中 */
  loading?: boolean;
  /** 额外的class名 */
  className?: string;
  /** 按钮active状态下额外的class名 */
  activeClassName?: string;
}
