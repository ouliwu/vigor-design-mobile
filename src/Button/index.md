---
title: Button
nav:
  title: 组件
  path: /components
---

## 基础使用

### type

通过 type 属性切换 Button 组件样式

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button>default</Button>
      <br />
      <Button type="primary">primary</Button>
      <br />
      <Button type="warn">warn</Button>
    </>
  );
};
```

### disabled

通过 disabled 属性设置按钮是否禁用

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button
        disabled={true}
        onPress={() => {
          console.log(1);
        }}
      >
        default
      </Button>
      <br />
      <Button disabled={true} type="primary">
        primary
      </Button>
      <br />
      <Button disabled={true} type="warn">
        warn
      </Button>
    </>
  );
};
```

<API></API>
