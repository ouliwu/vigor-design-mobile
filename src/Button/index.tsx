import React, { useState } from 'react';
import './index.scss';
import { ButtonPropsType } from './PropsType';
import Touchable from '@ant-design/mobile-touchable';
import classNames from 'classnames';

const prefixCls = 'h-button';

const Button: React.FC<ButtonPropsType> = props => {
  const {
    type,
    disabled,
    onPress,
    loading,
    className,
    activeClassName,
    children,
  } = props;
  const [cls, setCls] = useState({
    [prefixCls]: true,
    [`${prefixCls}-active`]: false,
    [activeClassName!]: false,
  });
  const innerCls = classNames({
    ...cls,
    [`${prefixCls}-${type}`]: true,
    [className!]: !!className,
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-loading`]: loading,
  });
  return (
    <Touchable
      disabled={loading || disabled}
      onPressIn={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: true,
          [activeClassName!]: true,
        });
      }}
      onPress={e => {
        onPress && onPress(e);
      }}
      onPressOut={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: false,
          [activeClassName!]: false,
        });
      }}
    >
      <div className={innerCls}>
        <div className={`${prefixCls}-loading-container`}>
          <div className={`${prefixCls}-loading-main`} />
        </div>
        <div className={`${prefixCls}-text-container`}>
          <span>{children}</span>
        </div>
      </div>
    </Touchable>
  );
};

Button.defaultProps = {
  type: 'default',
  disabled: false,
  loading: false,
  className: '',
  activeClassName: '',
};

export default Button;
