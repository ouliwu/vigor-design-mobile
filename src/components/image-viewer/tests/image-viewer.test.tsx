import React, { useRef, useState } from 'react'
import {
  render,
  testA11y,
  fireEvent,
  waitFor,
  screen,
  userEvent,
  mockDrag,
  act,
} from 'testing'
import ImageViewer, { MultiImageViewerRef } from '../index'
import Button from '../../button'

const classPrefix = `adm-image-viewer`

const demoImages = [
  'https://images.unsplash.com/photo-1620476214170-1d8080f65cdb?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3150&q=80',
  'https://images.unsplash.com/photo-1601128533718-374ffcca299b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3128&q=80',
  'https://images.unsplash.com/photo-1567945716310-4745a6b7844b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3113&q=80',
  'https://images.unsplash.com/photo-1624993590528-4ee743c9896e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=200&h=1000&q=80',
]

describe('ImageViewer', () => {
  test('a11y', async () => {
    await testA11y(<ImageViewer image={demoImages[0]} visible={true} />)
  })

  test('`ImageViewer.show/ImageViewer.clear` should be work', async () => {
    render(
      <button
        onClick={() => {
          ImageViewer.show({ image: demoImages[0] })
        }}
      >
        show
      </button>
    )
    fireEvent.click(screen.getByText('show'))
    const img = await screen.findByRole('img')
    expect(img).toBeVisible()

    act(() => {
      ImageViewer.clear()
    })
    await waitFor(() => expect(img).not.toBeVisible())
  })
})

describe('ImageViewer.Multi', () => {
  test('calling ref.current.swipeTo before initialization', async () => {
    function App() {
      const [visible, setVisible] = useState(false)
      const ref = useRef<MultiImageViewerRef>(null)
      return (
        <>
          <Button
            onClick={() => {
              ref.current?.swipeTo(2)
              setVisible(true)
            }}
          >
            Show
          </Button>
          <ImageViewer.Multi
            ref={ref}
            images={demoImages}
            visible={visible}
            defaultIndex={0}
          />
        </>
      )
    }
    const renderer = render(<App />)
    expect(renderer.container).toMatchSnapshot()
    fireEvent.click(renderer.getByText('Show'))
    await waitFor(() =>
      // end of animation
      expect(document.querySelectorAll('.adm-mask')[0]).toHaveStyle(
        'opacity: 1;'
      )
    )
    expect(renderer.getByText('3 / 4')).not.toBeNull()
    expect(renderer.container).toMatchSnapshot()
  })

  test('rendering with footer', () => {
    function App() {
      return (
        <ImageViewer.Multi
          images={demoImages}
          visible
          renderFooter={() => <Button>查看原图</Button>}
        />
      )
    }
    render(<App />)
    expect(screen.getByText('查看原图')).toBeInTheDocument()
  })

  test('`ImageViewer.Multi.show` should be work', async () => {
    render(
      <>
        <button
          onClick={() => {
            ImageViewer.Multi.show({ images: demoImages })
          }}
        >
          show
        </button>
      </>
    )
    fireEvent.click(screen.getByText('show'))
    const imgs = await screen.findAllByRole('img')
    expect(imgs[0]).toBeVisible()
    await userEvent.click(imgs[0])
    await waitFor(() => expect(imgs[0]).not.toBeVisible())
  })

  test('slide should be work', async () => {
    Object.defineProperty(window, 'innerWidth', {
      value: 300,
    })
    const onIndexChange = jest.fn()

    render(
      <button
        onClick={() => {
          ImageViewer.Multi.show({ images: demoImages, onIndexChange })
        }}
      >
        show
      </button>
    )

    fireEvent.click(screen.getByText('show'))
    await screen.findAllByRole('img')
    const slides = document.querySelectorAll(`.${classPrefix}-slides`)[0]
    expect(screen.getByText('1 / 4')).toBeInTheDocument()

    mockDrag(slides, [
      {
        clientX: 300,
      },
      {
        clientX: 200,
      },
      {
        clientX: 100,
      },
    ])
    await waitFor(() => expect(onIndexChange).toBeCalledWith(1))
    expect(screen.getByText('2 / 4')).toBeInTheDocument()
  })
})
