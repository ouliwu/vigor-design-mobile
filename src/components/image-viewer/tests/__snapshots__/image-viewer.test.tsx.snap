// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageViewer.Multi calling ref.current.swipeTo before initialization 1`] = `
<div>
  <button
    class="adm-button adm-button-default adm-button-shape-default"
    type="button"
  >
    <span>
      Show
    </span>
  </button>
</div>
`;

exports[`ImageViewer.Multi calling ref.current.swipeTo before initialization 2`] = `
<div>
  <button
    class="adm-button adm-button-default adm-button-shape-default"
    type="button"
  >
    <span>
      Show
    </span>
  </button>
  <div
    class="adm-mask"
    style="background: rgba(0, 0, 0, 0.75); opacity: 1;"
  >
    <div
      class="adm-mask-content"
    >
      <div
        class="adm-image-viewer-content"
      >
        <div
          class="adm-image-viewer-slides"
        >
          <div
            class="adm-image-viewer-indicator"
          >
            3 / 4
          </div>
          <div
            class="adm-image-viewer-slides-inner"
            style="transform: translate3d(-2080px,0,0);"
          >
            <div
              class="adm-image-viewer-slide"
            >
              <div
                class="adm-image-viewer-control"
              >
                <div
                  class="adm-image-viewer-image-wrapper"
                  style="transform: matrix(1,0,0,1,0,0);"
                >
                  <img
                    alt="https://images.unsplash.com/photo-1620476214170-1d8080f65cdb?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3150&q=80"
                    draggable="false"
                    src="https://images.unsplash.com/photo-1620476214170-1d8080f65cdb?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3150&q=80"
                  />
                </div>
              </div>
            </div>
            <div
              class="adm-image-viewer-slide"
            >
              <div
                class="adm-image-viewer-control"
              >
                <div
                  class="adm-image-viewer-image-wrapper"
                  style="transform: matrix(1,0,0,1,0,0);"
                >
                  <img
                    alt="https://images.unsplash.com/photo-1601128533718-374ffcca299b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3128&q=80"
                    draggable="false"
                    src="https://images.unsplash.com/photo-1601128533718-374ffcca299b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3128&q=80"
                  />
                </div>
              </div>
            </div>
            <div
              class="adm-image-viewer-slide"
            >
              <div
                class="adm-image-viewer-control"
              >
                <div
                  class="adm-image-viewer-image-wrapper"
                  style="transform: matrix(1,0,0,1,0,0);"
                >
                  <img
                    alt="https://images.unsplash.com/photo-1567945716310-4745a6b7844b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3113&q=80"
                    draggable="false"
                    src="https://images.unsplash.com/photo-1567945716310-4745a6b7844b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3113&q=80"
                  />
                </div>
              </div>
            </div>
            <div
              class="adm-image-viewer-slide"
            >
              <div
                class="adm-image-viewer-control"
              >
                <div
                  class="adm-image-viewer-image-wrapper"
                  style="transform: matrix(1,0,0,1,0,0);"
                >
                  <img
                    alt="https://images.unsplash.com/photo-1624993590528-4ee743c9896e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=200&h=1000&q=80"
                    draggable="false"
                    src="https://images.unsplash.com/photo-1624993590528-4ee743c9896e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=200&h=1000&q=80"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="adm-image-viewer-footer"
      >
        <div
          class="adm-safe-area adm-safe-area-position-bottom"
        />
      </div>
    </div>
  </div>
</div>
`;
